lib/go-atscfg/meta.go

```diff
@@ -0,0 +1,195 @@
+	}
+
+	atsData := tc.ATSConfigMetaData{
+		Info: tc.ATSConfigMetaDataInfo{
+			ProfileID:         int(server.ProfileID),
+			TOReverseProxyURL: tmReverseProxyURL,
+			TOURL:             tmURL,
+			ServerIPv4:        server.IP,
+			ServerPort:        server.Port,
+			ServerName:        server.HostName,
+			CDNID:             server.CDNID,
+			CDNName:           string(server.CDN),
+			ServerID:          server.ID,
+			ProfileName:       server.ProfileName,
+		},
+		ConfigFiles: []tc.ATSConfigMetaDataConfigFile{},
+	}
+
+	if locationParams["remap.config"].Location != "" {
+		configLocation := locationParams["remap.config"].Location
+		for _, ds := range uriSignedDSes {
+			cfgName := "uri_signing_" + string(ds) + ".config"
+			// If there's already a parameter for it, don't clobber it. The user may wish to override the location.
+			if _, ok := locationParams[cfgName]; !ok {
+				p := locationParams[cfgName]
+				p.FileNameOnDisk = cfgName
+				p.Location = configLocation
+				locationParams[cfgName] = p
+			}
+		}
+	}
+
+	for cfgFile, cfgParams := range locationParams {
+		atsCfg := tc.ATSConfigMetaDataConfigFile{
+			FileNameOnDisk: cfgParams.FileNameOnDisk,
+			Location:       cfgParams.Location,
+		}
+
+		scope := getServerScope(cfgFile, server.Type, scopeParams)
+
+		if cfgParams.URL != "" {
+			scope = tc.ATSConfigMetaDataConfigFileScopeCDNs
+			atsCfg.URL = cfgParams.URL
+		} else {
+			scopeID := ""
+			if scope == tc.ATSConfigMetaDataConfigFileScopeCDNs {
+				scopeID = string(server.CDN)
+			} else if scope == tc.ATSConfigMetaDataConfigFileScopeProfiles {
+				scopeID = server.ProfileName
+			} else { // ATSConfigMetaDataConfigFileScopeServers
+				scopeID = server.HostName
+			}
+			atsCfg.APIURI = "/api/1.2/" + string(scope) + "/" + scopeID + "/configfiles/ats/" + cfgFile
+		}
+
+		atsCfg.Scope = string(scope)
+
+		atsData.ConfigFiles = append(atsData.ConfigFiles, atsCfg)
+	}
+
+	bts, err := json.Marshal(atsData)
+	if err != nil {
+		// should never happen
+		log.Errorln("marshalling chkconfig NameVersions: " + err.Error())
+		bts = []byte("error encoding to json, see log for details")
+	}
+	return string(bts)
+}
+
+func getServerScope(cfgFile string, serverType string, scopeParams map[string]string) tc.ATSConfigMetaDataConfigFileScope {
+	switch {
+	case cfgFile == "cache.config" && tc.CacheTypeFromString(serverType) == tc.CacheTypeMid:
+		return tc.ATSConfigMetaDataConfigFileScopeServers
+	default:
+		return getScope(cfgFile, scopeParams)
+	}
+}
+
+const DefaultScope = tc.ATSConfigMetaDataConfigFileScopeServers
+
+// getScope returns the ATSConfigMetaDataConfigFileScope for the given config file, and potentially the given server. If the config is not a Server scope, i.e. was part of an endpoint which does not include a server name or id, the server may be nil.
+func getScope(cfgFile string, scopeParams map[string]string) tc.ATSConfigMetaDataConfigFileScope {
+	switch {
```

Comment:
Is the fallthrough here better than multiple-case? Like: `cfgFile == "ip_allow.config", cfgFile == "parent.config", ...`

Reply:
Changed.